tools/richlist.js

```diff
@@ -0,0 +1,559 @@
+var Block = require('../db.js').Block;
+
+mongoose.connect(process.env.MONGO_URI || 'mongodb://localhost/blockDB');
+
+const ADDRESS_CACHE_MAX = 10000; // address cache threshold
+
+function makeRichList(toBlock, blocks, updateCallback) {
+  var self = makeRichList;
+  if (!self.cached) {
+    self.cached = {};
+    self.index = 0;
+  }
+  if (!self.accounts) {
+    self.accounts = {};
+  }
+  var fromBlock = toBlock - blocks;
+  if (fromBlock < 0) {
+    fromBlock = 0;
+  }
+
+  console.log('Scan accounts from ' + fromBlock + ' to ' + toBlock + ' ...');
+
+  var ended = false;
+  if (fromBlock == toBlock) {
+    ended = true;
+  }
+
+  async.waterfall([
+    function(callback) {
+      // Transaction.distinct("from", { blockNumber: { $lte: toBlock, $gt: fromBlock } }, function(err, docs) ...
+      // faster
+      // dictint("from")
+      Transaction.aggregate([
+        { $match: { blockNumber: { $lte: toBlock, $gt: fromBlock } } },
+        { $group: { _id: '$from' }},
+        { $project: { "_id": 1 }}
+      ]).exec(function(err, docs) {
+        if (err) {
+          console.log(err);
+          return;
+        }
+        docs.forEach(function(doc) {
+          // check address cache
+          if (!self.cached[doc._id]) {
+            self.accounts[doc._id] = { address: doc._id, type: 0 };
+            // increase cache counter
+            self.cached[doc._id] = 1;
+          } else {
+            self.cached[doc._id]++;
+          }
+        });
+        callback(null);
+      });
+    }, function(callback) {
+      // dictint("to")
+      Transaction.aggregate([
+        { $match: { blockNumber: { $lte: toBlock, $gt: fromBlock } } },
+        { $group: { _id: '$to' }},
+        { $project: { "_id": 1 }}
+      ]).exec(function(err, docs) {
+        if (err) {
+           console.log(err);
+           return;
+        }
+        docs.forEach(function(doc) {
+          // to == null case
+          if (!doc._id) {
+            return;
+          }
+          if (!self.cached[doc._id]) {
+            self.accounts[doc._id] = { address: doc._id, type: 0 };
+            self.cached[doc._id] = 1;
+          } else {
+            self.cached[doc._id]++;
+          }
+        });
+        callback(null);
+      });
+    }, function(callback) {
+      // aggregate miner's addresses
+      Block.aggregate([
+        { $match: { number: { $lte: toBlock, $gt: fromBlock } } },
+        { $group: { _id: '$miner' }},
+        { $project: { "_id": 1 }}
+      ]).exec(function(err, docs) {
+        if (err) {
+           console.log(err);
+           return;
+        }
+        docs.forEach(function(doc) {
+          if (!self.cached[doc._id]) {
+            self.accounts[doc._id] = { address: doc._id, type: 0 };
+            self.cached[doc._id] = 1;
+          } else {
+            self.cached[doc._id]++;
+          }
+        });
+        callback(null);
+      });
+    }, function(callback) {
+      let len = Object.keys(self.accounts).length;
```

Comment:
What's the purpose of this function? To calculate total number of accounts?

Reply:
normal cases, tx size is small enough to call json-rpc batch in one time,
but for some other cases, total TX is too big, total accounts size too big to fit in one json-rpc batch call.
for these case, split accounts into chunks and call json-rpc batch job.

Reply:
Ok, makes sense here, resolving this part as I just wanted clarifications.